cmd/frontend/frontend.go

```diff
@@ -0,0 +1,172 @@
+package main
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"net"
+	"net/http"
+	"os"
+	"sync/atomic"
+
+	"github.com/Azure/ARO-HCP/pkg/api/arm"
+)
+
+const (
+	// Literal path segments must be lowercase because
+	// MiddlewareLowercase converts the request URL to
+	// lowercase before multiplexing.
+	ResourceProviderNamespace = "microsoft.redhatopenshift"
+
+	SubscriptionPath  = "/subscriptions/{subscriptionId}"
+	ResourceGroupPath = SubscriptionPath + "/resourcegroups/{resourceGroupName}"
+	ResourceTypePath  = ResourceGroupPath + "/" + ResourceProviderNamespace + "/{resourceType}"
+	ResourceNamePath  = ResourceTypePath + "/{resourceName}"
+)
+
+type PathValues struct {
+	SubscriptionID    string
+	ResourceGroupName string
+	ResourceType      string
+	ResourceName      string
+}
+
+func NewPathValues(request *http.Request) *PathValues {
+	return &PathValues{
+		SubscriptionID:    request.PathValue("subscriptionId"),
+		ResourceGroupName: request.PathValue("resourceGroupName"),
+		ResourceType:      request.PathValue("resourceType"),
+		ResourceName:      request.PathValue("resourceName"),
+	}
+}
+
+type Frontend struct {
+	logger   *slog.Logger
+	listener net.Listener
+	server   http.Server
+	ready    atomic.Value
+	done     chan struct{}
+}
+
+func NewFrontend(logger *slog.Logger, listener net.Listener) *Frontend {
+	f := &Frontend{
+		logger:   logger,
+		listener: listener,
+		server: http.Server{
+			ErrorLog: slog.NewLogLogger(logger.Handler(), slog.LevelError),
+		},
+		done: make(chan struct{}),
+	}
+
+	mux := NewMiddlewareMux(
+		MiddlewarePanic,
+		MiddlewareBody,
+		MiddlewareLowercase,
+		MiddlewareSystemData)
+
+	// Unauthenticated routes
+	mux.HandleFunc("/", f.NotFound)
+	mux.HandleFunc("GET /healthz/ready", f.HealthzReady)
+
+	// Authenticated routes
+	postMuxMiddleware := NewMiddleware(MiddlewareValidateAPIVersion)
+	mux.Handle(
+		"GET "+ResourceTypePath,
```

Comment:
for "GET", "PUT", ... should we make these constants?

Reply:
You mean like `http.MethodGet+" "+ResourceTypePath`?  Good idea.